File: api/v1/schemas/portfolio.py
# -*- coding: utf-8 -*-
"""
===================================
自选与交易档案 Schema
===================================
"""

from typing import List, Optional, Literal

from pydantic import BaseModel, Field


PortfolioStatus = Literal["holding", "watch", "candidate", "archived"]


class PortfolioProfileUpsertRequest(BaseModel):
    stock_code: str = Field(..., description="股票代码")
    stock_name: Optional[str] = Field(None, description="股票名称")
    status: PortfolioStatus = Field("watch", description="状态")
    is_favorite: bool = Field(False, description="是否收藏关注")

    buy_price: Optional[float] = Field(None, ge=0, description="买入价")
    position_pct: Optional[float] = Field(None, ge=0, le=100, description="仓位百分比")
    shares: Optional[float] = Field(None, ge=0, description="持仓股数")
    target_buy_price: Optional[float] = Field(None, ge=0, description="目标入场价")
    target_sell_price: Optional[float] = Field(None, ge=0, description="目标止盈价")
    stop_loss_price: Optional[float] = Field(None, ge=0, description="止损价")

    tags: List[str] = Field(default_factory=list, description="标签列表")
    notes: Optional[str] = Field(None, description="备注")


class PortfolioProfile(BaseModel):
    id: int
    stock_code: str
    stock_name: Optional[str] = None
    status: PortfolioStatus
    is_favorite: bool = False

    buy_price: Optional[float] = None
    position_pct: Optional[float] = None
    shares: Optional[float] = None
    target_buy_price: Optional[float] = None
    target_sell_price: Optional[float] = None
    stop_loss_price: Optional[float] = None
    tags: List[str] = Field(default_factory=list)
    notes: Optional[str] = None

    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class PortfolioProfileListResponse(BaseModel):
    total: int
    items: List[PortfolioProfile]

File: apps/dsa-web/src/types/portfolio.ts
export type PortfolioStatus = 'holding' | 'watch' | 'candidate' | 'archived';

export interface PortfolioProfile {
  id: number;
  stockCode: string;
  stockName?: string;
  status: PortfolioStatus;
  isFavorite: boolean;
  buyPrice?: number;
  positionPct?: number;
  shares?: number;
  totalInvestment?: number;
  targetBuyPrice?: number;
  targetSellPrice?: number;
  stopLossPrice?: number;
  tags: string[];
  actionHistory?: string[];
  notes?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface PortfolioProfileListResponse {
  total: number;
  items: PortfolioProfile[];
}

export interface PortfolioUpsertRequest {
  stockCode: string;
  stockName?: string;
  status: PortfolioStatus;
  isFavorite: boolean;
  buyPrice?: number;
  positionPct?: number;
  shares?: number;
  totalInvestment?: number;
  targetBuyPrice?: number;
  targetSellPrice?: number;
  stopLossPrice?: number;
  tags?: string[];
  actionHistory?: string[];
  notes?: string;
}

export interface PortfolioReviewResult {
  availableCash: number;
  holdingsCount: number;
  holdings?: Array<{
    stockCode: string;
    stockName?: string;
    positionPct?: number;
    latestScore?: number;
    latestAdvice?: string;
  }>;
  riskDiversification?: {
    industryConcentration?: string;
    topIndustryExposurePct?: number;
    assessment?: string;
  };
  positionRecommendation?: {
    marketFearGreed?: {
      score?: number;
      label?: string;
    };
    totalPositionPct?: number;
    assessment?: string;
    suggestedRangePct?: string;
  };
  bulletPlan?: {
    buyList?: Array<{
      stockCode: string;
      stockName?: string;
      sectorName?: string;
      latestScore?: number;
      reason?: string;
    }>;
    availableCashUsageHint?: string;
  };
  message?: string;
}
